src/visitors/hoist.js

```diff
@@ -183,122 +182,201 @@ module.exports = {
+          getName(asset, 'require', path.node.source.value)
+        );
       }
     }
+
+    path.remove();
   },
-  ExportNamedDeclaration(path, asset) {
-    let {declaration, source, specifiers} = path.node;
 
-    if (!source) {
-      let declarations;
+  ExportDefaultDeclaration(path, asset) {
+    let {declaration} = path.node;
+    let identifier = getIdentifier(asset, 'export', 'default');
 
-      if (declaration) {
-        declarations = declaration.declarations.map(decl => {
-          asset.cacheData.exports.push(decl.id.name);
+    if (t.isExpression(declaration)) {
+      // Declare a variable to hold the exported value.
+      path.replaceWith(
+        t.variableDeclaration('var', [
+          t.variableDeclarator(identifier, declaration)
+        ])
+      );
+    } else {
+      // Rename the declaration to the exported name.
+      path.replaceWith(declaration);
+      path.scope.rename(declaration.id.name, identifier.name);
+    }
 
-          return getNamedExportVarDecl(asset, decl.id, decl.init);
-        });
-      } else if (specifiers.length > 0) {
-        declarations = t.variableDeclaration(
-          'var',
-          specifiers.map(specifier => {
-            asset.cacheData.exports.push(specifier.exported.name);
+    // Add assignment to exports object for namespace imports and commonjs.
+    path.insertAfter(
+      EXPORT_ASSIGN_TEMPLATE({
+        EXPORTS: getExportsIdentifier(asset),
+        NAME: t.identifier('default'),
+        LOCAL: identifier
+      })
+    );
+
+    asset.cacheData.exports[identifier.name] = 'default';
+
+    // Mark the asset as an ES6 module, so we handle imports correctly in the packager.
+    asset.cacheData.isES6Module = true;
+  },
+
+  ExportNamedDeclaration(path, asset) {
+    let {declaration, source, specifiers} = path.node;
+
+    if (source) {
+      for (let specifier of specifiers) {
+        // Create a variable to re-export from the imported module.
+        path.insertAfter(
+          t.variableDeclaration('var', [
+            t.variableDeclarator(
+              getIdentifier(asset, 'export', specifier.exported.name),
+              getIdentifier(asset, 'import', source.value, specifier.local.name)
+            )
+          ])
+        );
 
-            return getNamedExportVarDecl(
+        path.insertAfter(
+          EXPORT_ASSIGN_TEMPLATE({
+            EXPORTS: getExportsIdentifier(asset),
+            NAME: t.identifier(specifier.exported.name),
+            LOCAL: getIdentifier(
               asset,
-              specifier.exported,
-              specifier.local
-            );
+              'import',
+              source.value,
+              specifier.local.name
+            )
           })
         );
       }
 
-      if (declarations.length) {
-        path.replaceWith(t.variableDeclaration('var', declarations));
+      path.remove();
+    } else if (declaration) {
+      path.replaceWith(declaration);
+
+      if (
```

Comment:
Added support for a bunch of different types of declarations. Still more to support, see todos.